File: week 3/dot_product/dot_product.cpp
#include <algorithm>
#include <iostream>
#include <vector>
#include <cassert>

using std::vector;

void selection_sort(vector<int> &a) {
  size_t size = a.size();
  int max_index, temp;

  for (int i = 0; i < size - 1; i++) {
    max_index = i;
    for (int j = i + 1; j < size; j++) {
      if (a[j] > a[max_index]) {
        max_index = j;
      }
    }

    temp = a[i];
    a[i] = a[max_index];
    a[max_index] = temp;
  }
}

long long max_dot_product(vector<int> a, vector<int> b) {
  // sort
  selection_sort(a);
  selection_sort(b);

  long long result = 0;
  for (size_t i = 0; i < a.size(); i++) {
    std::cout << a[i] << ' ' << b[i] << '\n';
    result += ((long long) a[i]) * b[i];
  }

  std::cout << result << '\n';
  return result;
}

void test_solution() {
  assert(max_dot_product({23}, {39}) == 897);
  assert(max_dot_product({1, 3, -5}, {-2, 4, 1}) == 23);
  assert(max_dot_product({3, 2, 8}, {-2, -3, -4}) == -33);
}

int main() {
  test_solution();

  size_t n;
  std::cin >> n;
  vector<int> a(n), b(n);
  for (size_t i = 0; i < n; i++) {
    std::cin >> a[i];
  }
  for (size_t i = 0; i < n; i++) {
    std::cin >> b[i];
  }
  std::cout << max_dot_product(a, b) << std::endl;
}

File: week 3/dot_product/dot_product.py
#Uses python3
import sys
import math

def selection_sort(a):
    for i in range(0, len(a) - 1):
        curr_index = i
        for j in range(i + 1, len(a)):
            if a[j] > a[curr_index]:
                curr_index = j

        a[i], a[curr_index] = a[curr_index], a[i]

    return a

def max_dot_product(a, b):
    assert len(a) == len(b)

    a = selection_sort(a)
    b = selection_sort(b)

    res = 0

    for i in range(len(a)):
        res += a[i] * b[i]

    return res

def test_solution():
    assert max_dot_product([23], [39]) == 897
    assert max_dot_product([1, 3, -5], [-2, 4, 1]) == 23
    assert max_dot_product([3, 2, 8], [-2, -3, -4]) == -33

if __name__ == '__main__':
    test_solution()

    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n = data[0]
    a = data[1:(n + 1)]
    b = data[(n + 1):]
    print(max_dot_product(a, b))
